src/store/modules/web3-providers.module.ts

```diff
@@ -138,6 +158,103 @@ export const useWeb3ProvidersStore = defineStore(STORE_NAME, () => {
     await sleep(1000)
   }
 
+  const getUsersLastFullyDeployedProtocol = async (): Promise<void> => {
+    const [l1DeployedPools, l2DeployedPools] = await Promise.all([
+      l1FactoryContract.value.providerBased.value.getDeployedPools(
+        String(_route.query.address),
+        0,
+        MAX_UINT_256,
+      ),
+      l2FactoryContract.value.providerBased.value.getDeployedPools(
+        String(_route.query.address),
+        0,
+        MAX_UINT_256,
+      ),
+    ])
+
+    let lastFullyL1DeployedPoolIdx = -1
+    let lastFullyL2DeployedPoolIdx = -1
+    for (let i = l1DeployedPools.length - 1; i >= 0; i--) {
+      if (
+        lastFullyL1DeployedPoolIdx !== -1 &&
+        lastFullyL2DeployedPoolIdx !== -1
+      )
+        break
+
+      for (let j = l2DeployedPools.length - 1; j >= 0; j--) {
+        if (l1DeployedPools[i].protocol === l2DeployedPools[j].protocol) {
+          lastFullyL1DeployedPoolIdx = i
+          lastFullyL2DeployedPoolIdx = j
+          break
+        }
+      }
+    }
+
+    if (
+      lastFullyL1DeployedPoolIdx === -1 ||
+      lastFullyL2DeployedPoolIdx === -1
+    ) {
+      return
+    }
+
+    dashboardInfo.name = l1DeployedPools[lastFullyL1DeployedPoolIdx].protocol
+    dashboardInfo.distributionAddress =
+      l1DeployedPools[lastFullyL1DeployedPoolIdx].distribution
+    dashboardInfo.l1SenderAddress =
+      l1DeployedPools[lastFullyL1DeployedPoolIdx].l1Sender
+    dashboardInfo.l2MessageReceiverAddress =
+      l2DeployedPools[lastFullyL2DeployedPoolIdx].l2MessageReceiver
+    dashboardInfo.l2TokenReceiverAddress =
+      l2DeployedPools[lastFullyL2DeployedPoolIdx].l2TokenReceiver
+    dashboardInfo.tokenAddress =
+      l2DeployedPools[lastFullyL2DeployedPoolIdx].mor20
+  }
+
+  watch(
+    () => _route.query,
+    () => {
+      if (_route.query.address) {
+        getUsersLastFullyDeployedProtocol()
+        return
+      }
+      dashboardInfo.name = ''
+      dashboardInfo.distributionAddress = ''
+      dashboardInfo.l1SenderAddress = ''
+      dashboardInfo.l2MessageReceiverAddress = ''
+      dashboardInfo.l2TokenReceiverAddress = ''
+      dashboardInfo.tokenAddress = ''
+    },
+    {
+      immediate: true,
+    },
+  )
+
+  watch(
+    dashboardInfo,
+    async () => {
+      if (!dashboardInfo.tokenAddress) {
+        dashboardDepositTokenAddress.value = ''
+        dashboardRewardTokenAddress.value = ''
+        return
+      }
+      dashboardDepositTokenAddress.value =
+        await erc1967ProxyContract.value.providerBased.value.depositToken()
+      dashboardRewardTokenAddress.value = dashboardInfo.tokenAddress
+    },
+    { deep: true },
+  )
```

Comment:
Seems like too much `dashboardInfo`  related logic for `web3-providers.module`.

Move it to separate composable to make this module cleaner and easier to maintain.

**Also optimization tip.**
Instead of looping 2 nested loops you can use `Map` for quick access and use single loop that will **reduce the time complexity** from O(n^2) to O(n)

Something like this:
```ts
const l2DeployedPoolMap = new Map(l2DeployedPools.map(pool => [pool.protocol, pool]));

  for (let i = l1DeployedPools.length - 1; i >= 0; i--) {
    const l1Pool = l1DeployedPools[i];
    const l2Pool = l2DeployedPoolMap.get(l1Pool.protocol);

    if (l2Pool) {
      dashboardInfo.name = l1Pool.protocol;
      dashboardInfo.distributionAddress = l1Pool.distribution;
      dashboardInfo.l1SenderAddress = l1Pool.l1Sender;
      dashboardInfo.l2MessageReceiverAddress = l2Pool.l2MessageReceiver;
      dashboardInfo.l2TokenReceiverAddress = l2Pool.l2TokenReceiver;
      dashboardInfo.tokenAddress = l2Pool.mor20;
      break;
    }
  }
```
Make sure that this snippet produces the same result, I didn't test it)

Reply:
it's thing also with TODO

Reply:
Okay, leave decomposing logic to hooks for later. At least apply optimization to nested loop algorithm.